### ISSUE TITLE
Mouse cursor gets stuck and locked in the viewport when tweaking sliders

### ISSUE THREAD

**Issue by Rabocop** [godotengine/godot#11893]
Operating system : Windows 10 pro 64 bits .
Godot version : All the 3.0 builds,
GPU : GeForce GTX 1060 6gb
Driver version : 382.05
**Issue description:**
![grut](https://user-images.githubusercontent.com/7796648/31300827-d4312ce2-aaed-11e7-9498-0bcdea88b733.gif)

When trying to tweak sliders, the cursor dissapears and gets stuck in the viewport, which turn the engine practically unusable. 
I found an issue long ago and I thought it was solved, and I can't find it.

IMPORTANT : It only happens in Windows (and I think it happened in my previus pc, with an AMD 7950 and Windows 7). On Linux I'm not having that problem

**Steps to reproduce:**
Tweak values until it happens.

**Comment by fracteed** [comment]
I can also confirm that I have had this happen randomly for a long time with Win10. I use a Wacom tablet almost exclusively, so I always thought it may have been related. Are you using a mouse or trackpad? When this does happen for me (usually when tweaking a slider in Spatial Material), I have to shutdown Godot via ctrl/alt/delete as my cursor entirely disappears from the OS.

Good to see that you have added this as an issue to github, as it is a very hard bug to replicate!

**Comment by Zylann** [comment]
I remember a similar issue was reported before, but I don't know which (hard to find from my phone)

**Comment by Rabocop** [comment]
@fracteed I messed a lot and found a way to get the cursor back, ctrl+A and add whatever node, then ctrl+z and then you have it again, it's annoying and you lose the focus in the inspector, but at least is something.

**Comment by fracteed** [comment]
@Rabocop thanks I will keep that in mind. Hopefully that workaround may help to isolate the root cause of this issue. I am now in a habit of just typing in values, as opposed to using the slider, so it has been a while since I have run into this bug!

**Comment by akien-mga** [comment]
Is this still valid in the current master branch?

**Comment by fracteed** [comment]
@akien-mga  Yes, this is still a current problem. I actually avoid using the sliders, as this is when it happens the most, but it does also happen less frequently with general interaction in the inspector.

I do use an Intuos tablet almost exclusively, so not sure if this exacerbates the problem or not? Note that I am using a current build under Win10.

**Comment by reduz** [comment]
I cant reproduce this, guess will have to be kicked to 3.1 and expect better reproduction steps or hardware or someone who can take a look at it who can reproduce

**Comment by bojidar-bg** [comment]
Perhaps it is related to the implementation of `MOUSE_MODE_CONFINED` on the windows platform, not sure...

**Comment by ILoveGodot** [comment]
It happens on Linux as well. I'm using Kubuntu 17.10 with a GP107M (GeForce GTX 1050 Ti Mobile, driver version: 384.111) with Godot 3.0. But it only happens to me when I try to change the anchor properties (Left, Top, Right, Bottom) of a control node in the inspector. Only then it warps the mouse position to the center of the viewport. Sometimes I manage to increase the slider value by a small fraction or set it to the max value. Changing the margin properties right below the anchor category seems to work fine.

**Comment by tsheilds7** [comment]
Happened to me too on Windows 10

**Comment by SojanJanso** [comment]
I can confirm @ILoveGodot mention of it only happening on the anchor properties. For me it is not locking it, but it is definitely not working like the other sliders. My mouse is not locking down in the middle, but it is warping there for a fraction of a second, wich can be seen in the giff if you look close enough. Just thought it could help narrow down the issue. I'm on Ubuntu 17.10 using a GTX 780 drivers 384.111.

![slider](https://user-images.githubusercontent.com/26679320/37735258-ea3ae65a-2d23-11e8-843c-32d5d7bb3f49.gif)

**Comment by Zylann** [comment]
It happened to me in 3.0.2 as I was tweaking properties of a tool node I made. It was working normally, then on some random property it started locking at the center of the screen, and now doing that for any property.

**Comment by YaharaOctanis** [comment]
In 3.0.2 on Windows 10 I found a reliable way of reproducing the bug, but it's pretty bizarre.

1. Create and open new project
2. Maximize the editor
3. Load default godot editor layout
4. On the right, increase the vertical height of inspector (shrink the scene view) as much as possible. Make sure it is above the center of the screen.
5. Add Particles2D node to the scene
6. In 'Process Material' create new 'ParticlesMaterial'
7. Expand 'Gravity', 'Initial Velocity' and 'Angular Velocity' menus
8. In inspector click and drag from lower left corner of 'Velocity random' value in 'Initial Velocity' menu.

![lowerleft_corner](https://user-images.githubusercontent.com/23387327/39590943-9235b7d6-4f02-11e8-9e70-e7567d2b5ab8.png)


It only worked when the value I tried to drag was above the center of the screen. I was able to reproduce it 3 times following the steps I included above. How or why it happens, I am not sure, but I hope this helps.

**Comment by Zylann** [comment]
@YaharaOctanis is the viewport getting focus when you do this?

**Comment by Zylann** [comment]
The viewport should not get focus, even if the inspector switches to warped cursor mode. The focus should be left to the spinbox until dragging is finished and the mouse returns to its initial state.
I think RMB in viewport is a way to unlock the mouse because it is a shorthand for fly mode, when the viewport is focused.

**Comment by Oblugi** [comment]
Removing this line seems to fix the issue for the anchor properties of control nodes that @ILoveGodot mentioned:
https://github.com/godotengine/godot/blob/af9a6202ebb1eda4f928d2d601162cf4b211752b/scene/gui/control.cpp#L1399
I think this prevents calls to the `update_tree()` function of the `PropertyEditor` class. The function basically refreshes all properties in the inspector which seems to mess up some UI interactions like dragging a value with the mouse cursor or dragging the horizontal slider to change a value.

This issue seems to be related:
https://github.com/godotengine/godot/issues/9890

**Comment by Zylann** [comment]
That means we would have to wait for inspector rewrite, because it seems `_change_notify` is needed (in case you modify a property but that property performs clamping or validation for example). It should have little to do with focus handling.

**Comment by Jose-Moreno** [comment]
I can confirm this is happening in **3.0.4** Steam version. I can reproduce it 100% of the time. In my example test I used a Button Node.

1. Left click and drag on a value slider and while dragging a value press right click. 
2. The cursor will lock in the middle of the screen. 

The only **workaround** I've found is:

1. Pres the TAB key several times until you shift the focus on the inspector
2. Then use the UP arrow key to "get out" of the slider focus lock (you have to press UP until you get out of the inspector though)

**Comment by Zylann** [comment]
I think that bug went away in master, but I need to recheck

**Comment by Jose-Moreno** [comment]
@Zylann I can't try master right now, but i'll try again later with 3.0.5 (or i'll try compiling master) to confirm your test. Thanks for taking the time to look into this as well :slightly_smiling_face:

**Comment by RodZill4** [comment]
I have a similar problem in v3.0.6 when, during a drag'n'drop operation, the cursor touches the (up/down) arrows of a SpinBox: the mouse cursor is captured at the center of the window.

**Comment by Wiadomy** [comment]
Same problem here, on Linux.
Cursors becomes invisible after changing some slider values.
Version 3.1.dev-1

**Comment by AlbertMontagutCasero** [comment]
I have the same issue in the 3.0.6 too. when i add sprites in AnimatedSprite my mouse goes to center of the screen and locked. I need to reset the entire editor to fix.

**Comment by Zylann** [comment]
@AdamRatai @AlbertMontagutCasero did you try the master branch or the most recent alpha? I think it got fixed there instead.

**Comment by reduz** [comment]
Again, please add version and reproduction conditions for this, as well as operating system, else your comments are not useful.

**Comment by AdamRatai** [comment]
Greets, mr Juan!  I have this in 3.0.6 stable official, in few different material sliders. I think it didn't happen as long as I don't drag in a input cell. Windows 10, nvidia GTX 10xx. If I only find a steps to reproduce, I will post asap.  Edit material, drag to change value in input done.  I not seeing this in 3.1 at all, I will check 3.0.6 once again.

**Comment by reduz** [comment]
Oh yeah, glad to know then, yes it should be fixed in 3.1

**Comment by raelik** [comment]
Yeah, this happens to me whenever I try to drag a resource into the frame list for an animation... which happens to make this app completely unusable for me.

**Comment by psionic12** [comment]
same as @raelik, I don't know if this has something to do with platform, mine is mac os.

**Comment by gerrywastaken** [comment]
This issue is happening frequently for me on the latest alpha of version Godot v4. I've figured out how to reproduce it fairly consistently and simply and created a new issue as was requested by @reduz earler:
https://github.com/godotengine/godot/issues/11893#issuecomment-418886151
